src/main/java/whisper/Command.java

```diff
@@ -0,0 +1,147 @@
+package whisper;
+import java.util.ArrayList;
+
+public interface Command {
+    void execute(ArrayList<Task> tasks, Ui ui, Storage storage) throws WhisperException;
+    boolean isExit();
+}
+
+// Add Command
+class AddCommand implements Command {
+    private Task taskToAdd;
+
+    public AddCommand(Task taskToAdd) {
+        this.taskToAdd = taskToAdd;
+    }
+
+    @Override
+    public void execute(ArrayList<Task> tasks, Ui ui, Storage storage) {
+        try {
+            // Add the task directly to the original tasks list
+            tasks.add(taskToAdd);
+            // Save the updated tasks list to the file
+            storage.saveFile(tasks);
+            ui.printTaskAdded(taskToAdd, tasks.size());
+        } catch (WhisperException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public boolean isExit() {
+        return false;
+    }
+}
+
+// Delete Command
+class DeleteCommand implements Command {
+    private int taskIndex;
+
+    public DeleteCommand(int taskIndex) {
+        this.taskIndex = taskIndex;
+    }
+
+    @Override
+    public void execute(ArrayList<Task> tasks, Ui ui, Storage storage) throws WhisperException {
+        try {
+            if (taskIndex >= 0 && taskIndex < tasks.size()) {
+                Task removedTask = tasks.remove(taskIndex);
+                ui.printRemovedTask(removedTask, tasks.size());
+                storage.saveFile(tasks);
+            } else {
+                throw WhisperException.invalidTaskID();
+            }
+        } catch (WhisperException e) {
+            throw new WhisperException("Invalid task number.");
+        }
+    }
+
+    @Override
+    public boolean isExit() {
+        return false;
+    }
+}
+
+// Exit Command
+class ExitCommand implements Command {
+    @Override
+    public void execute(ArrayList<Task> tasks, Ui ui, Storage storage) {
+        ui.printExitMessage();
+    }
+
+    @Override
+    public boolean isExit() {
+        return true;
+    }
+}
+
+// List Command
+class ListCommand implements Command {
+    @Override
+    public void execute(ArrayList<Task> tasks, Ui ui, Storage storage) {
+        ui.printTasks(tasks);
+    }
+
+    @Override
+    public boolean isExit() {
+        return false;
+    }
+}
+
+// Mark Command
+class MarkCommand implements Command {
+    private int taskIndex;
+
+    public MarkCommand(int taskIndex) {
+        this.taskIndex = taskIndex;
+    }
+
+    @Override
+    public void execute(ArrayList<Task> tasks, Ui ui, Storage storage) throws WhisperException {
+        try {
+            if (taskIndex >= 0 && taskIndex < tasks.size()) {
+                tasks.get(taskIndex).markAsDone();
+                ui.printTaskAsDone(tasks.get(taskIndex));
+                storage.saveFile(tasks);
+            } else {
+                ui.printError("Invalid task number. Please choose between task 1 - " + tasks.size() + "\n");
```

Comment:
In the MarkCommand and UnmarkCommand classes, the error message is directly printed to the UI. It might be better to throw exceptions and handle them in a higher level of the application, leaving UI classes for UI-related tasks only 👍

Reply:
Thanks for the suggestion!